### PR TITLE
IGNITE-20381 Prevent modification operations call after CatalogManagerImpl is closed

### PR DESCRIPTION
**Pull request by zstan**
…rImpl is closed

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java [hunk 1]
```diff
@@ -321,8 +328,10 @@ public CompletableFuture<Void> execute(List<CatalogCommand> commands) {
     }
 
     private void registerCatalog(Catalog newCatalog) {
-        catalogByVer.put(newCatalog.version(), newCatalog);
-        catalogByTs.put(newCatalog.time(), newCatalog);
+        inBusyLock(busyLock, () -> {
+            catalogByVer.put(newCatalog.version(), newCatalog);
+            catalogByTs.put(newCatalog.time(), newCatalog);
+        });
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java:334]: This method is called in `OnUpdateHandlerImpl.handle(...)`.
` handleFutures.add(onUpdateHandler.handle(update, event.timestamp(), event.revision()));`
It make sense to guard the `UpdateListener.onUpdate(WatchEvent)` with busylock instead.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java [hunk 2]
```diff
@@ -170,6 +176,7 @@ public void start() {
 
     @Override
     public void stop() throws Exception {
+        busyLock.block();
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java:179]: Should we close versionTracker to unblock all the waiters?

Comment (AMashenkov) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java:179]: And is it ok to allow getting latest catalog version after CatalogManager was stopped?

We may violate invariant and the one will wait for a new catalog version, but the new version event will never be accepted due to stopping.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java (hunk 1 above)
Comment (zstan) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java]: seems no need to use busyLock here it correctly stops before.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java (hunk 2 above)
Comment (zstan) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java:178]: done, tests fixed

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**